src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,6 +264,35 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
+							if (getParser().getCurrentStructure() != null) {
+								if (t.getUsableStructures().length > 0) {
+									boolean error = true;
+									Class<?> current = getParser().getCurrentStructure().getClass();
+									for (Class<? extends Structure> clazz : t.getUsableStructures()) {
+										if (clazz.isAssignableFrom(current))
+											error = false;
+									}
+									if (error)
+										Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(getParser().getCurrentStructure().toString(null, false)) + " structure.");
+								}
+							} else if (t.getUsableStructures().length > 0) {
+								Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+							}
+							List<TriggerSection> sections = getParser().getCurrentSections();
+							TriggerSection current = sections.isEmpty() ? null : sections.get(sections.size() - 1);
+							if (t.getUsableSections().length > 0) {
+								if (current == null) {
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed sections.");
+								} else {
+									boolean error = true;
+									for (Class<? extends Section> clazz : t.getUsableSections()) {
+										if (clazz.isAssignableFrom(current.getClass()))
+											error = false;
+									}
+									if (error)
+										Skript.error("'" + res.expr + "' cannot be used in a '" + current.toString(null, false) + "' section.");
+								}
+							}
```

Comment:
two things:
1) You could probably use ParserInstance#isCurrentStructure and ParserInstance#IsCurrentSection instead of manual checks :D 
2) Do you need to return (null?) if the error is printed? I don't think we want `init` to be called.

Reply:
1. You are most likely right, it seems to do the same thing.
2. You make a great point, most likely not, i'll add these in now c: